axum-jwks/src/key_manager.rs

```diff
@@ -0,0 +1,143 @@
+use std::sync::Arc;
+
+use crate::key_store::JwksError;
+use crate::{KeyStore, TokenError};
+use jsonwebtoken::{decode, decode_header, TokenData};
+use serde::de::DeserializeOwned;
+
+use tokio::sync::RwLock;
+
+use tokio::time::{sleep, Duration, Instant};
+use tracing::{debug, error, info};
+
+#[derive(Clone)]
+pub struct KeyManager {
+    authority: String,
+    audience: String,
+    minimal_interval: Option<Duration>,
+    key_store: Arc<RwLock<KeyStore>>,
+    client: reqwest::Client,
+}
+
+impl KeyManager {
+    /// Create a new KeyManager that can fetch jwks from an authority
+    /// `authority`: either url of an openid_configuration or a jwks_url
+    /// `audience`: to be checked against the `aud` claim
+    ///
+    /// jwks is not initially fetched: Please see `update_now` function
+    pub fn new(authority: String, audience: String) -> Self {
+        Self {
+            authority,
+            audience,
+            minimal_interval: None,
+            key_store: Arc::new(RwLock::new(KeyStore::default())),
+            client: reqwest::Client::default(),
+        }
+    }
+
+    /// When validating a token: fetch updated jwks if kid not found
+    ///
+    /// Do not update more often than `interval`
+    pub fn with_minimal_update_interval(mut self, interval: u64) -> Self {
+        self.minimal_interval = Some(Duration::from_secs(interval));
+        self
+    }
+
+    /// Periodically update the jwks every `interval`, including immediately
+    pub fn with_periodical_update(self, interval: u64) -> Self {
+        let key_store = self.key_store.clone();
+        let url = self.authority.clone();
+        let audience = self.audience.clone();
+        let client = self.client.clone();
+        tokio::spawn(async move {
+            let duration = Duration::from_secs(interval);
+            loop {
+                {
+                    let mut ks = key_store.write().await;
+                    match KeyStore::new(&client, &url, &audience).await {
+                        Ok(new_ks) => {
+                            info!("Periodically updated jwks");
+                            *ks = new_ks
+                        }
+                        Err(e) => error!(?e, "Could not update jwks from: {}", url),
```

Comment:
I think this error needs to be bubbled up to the caller instead of just being logged. As a consumer of the library, I would at least like the option to choose how to handle errors.